src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineHelper.cs

```diff
@@ -0,0 +1,340 @@
+                    currentGroup = ArrayBuilder<DocumentSymbol>.GetInstance(1, symbol);
+                    currentRange = symbol.Range;
+                }
+            }
+
+            documentSymbolGroups.Add(currentGroup.ToImmutableAndFree());
+            return documentSymbolGroups.ToImmutableAndFree();
+        }
+
+        /// <summary>
+        /// Given a flat array containing a Document Symbol and its descendants, returns the Document Symbol
+        /// with its descendants recursively nested
+        /// </summary>
+        /// /// <remarks>
+        /// The first Document Symbol in the array is considered the root node.
+        /// </remarks>
+        private static DocumentSymbol CreateDocumentSymbolTree(ImmutableArray<DocumentSymbol> documentSymbols)
+        {
+            var node = documentSymbols.First();
+            var childDocumentSymbols = documentSymbols.RemoveAt(0);
+            node.Children = GroupDocumentSymbolTrees(childDocumentSymbols)
+                .Select(group => CreateDocumentSymbolTree(group))
+                .ToArray();
+            return node;
+        }
+
+        /// <summary>
+        /// Converts an array of type DocumentSymbol to an array of type DocumentSymbolViewModel
+        /// </summary>
+        public static ImmutableArray<DocumentSymbolViewModel> GetDocumentSymbolModels(DocumentSymbol[] documentSymbols)
+        {
+            var documentSymbolModels = ArrayBuilder<DocumentSymbolViewModel>.GetInstance();
+            if (documentSymbols.Length == 0)
+            {
+                return documentSymbolModels.ToImmutableAndFree();
+            }
+
+            foreach (var documentSymbol in documentSymbols)
+            {
+                var documentSymbolModel = new DocumentSymbolViewModel(documentSymbol);
+                if (documentSymbol.Children is not null)
+                    documentSymbolModel.Children = GetDocumentSymbolModels(documentSymbol.Children);
+                documentSymbolModels.Add(documentSymbolModel);
+            }
+
+            return documentSymbolModels.ToImmutableAndFree();
+        }
+
+        public static async Task<ManualInvocationResponse?> DocumentSymbolsRequestAsync(
+            ITextBuffer textBuffer,
+            ILanguageServiceBroker2 languageServiceBroker,
+            string? textViewFilePath,
+            CancellationToken cancellationToken)
+        {
+            if (textViewFilePath is null)
+                return null;
+
+            var parameterFactory = new DocumentSymbolParams()
+            {
+                TextDocument = new TextDocumentIdentifier()
+                {
+                    Uri = new Uri(textViewFilePath)
+                }
+            };
+
+            // TODO: proper workaround such that context.ClientCapabilities?.TextDocument?.DocumentSymbol?.HierarchicalDocumentSymbolSupport == true
+            return await languageServiceBroker.RequestAsync(
+                textBuffer: textBuffer,
```

Comment:
odd that the API requires the textBuffer _and_ the path to the file. I would expect it to just need one? @dibarbet do you know the reason the API is shaped this way?

@emilyanas2323 there is nothing you need to change here; I am just surprised this is the API the editor gives us.

Reply:
No real reason afaik, just an old API that needs updating.  Tracking issue
https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1542016/